File: sdk/python/aragora/namespaces/email_services.py
"""
Email Services Namespace API

Provides methods for advanced email management services including
follow-up tracking, snooze management, and email categorization.

Features:
- Follow-up tracking (mark, list, resolve, check replies)
- Snooze recommendations and management
- Email categorization
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class EmailServicesAPI:
    """
    Synchronous Email Services API.

    Provides methods for advanced email management:
    - Follow-up tracking
    - Snooze management
    - Email categorization

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="...")
        >>> pending = client.email_services.list_pending_followups()
        >>> client.email_services.mark_followup(email_id="abc123")
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # ===========================================================================
    # Follow-up Tracking
    # ===========================================================================

    def mark_followup(
        self,
        email_id: str,
        expected_reply_by: str | None = None,
        notes: str | None = None,
    ) -> dict[str, Any]:
        """
        Mark an email as awaiting reply.

        Args:
            email_id: ID of the email to track
            expected_reply_by: Optional ISO datetime for expected reply
            notes: Optional notes about the follow-up

        Returns:
            Dict with followup_id and status
        """
        data: dict[str, Any] = {"email_id": email_id}
        if expected_reply_by:
            data["expected_reply_by"] = expected_reply_by
        if notes:
            data["notes"] = notes

        return self._client.request("POST", "/api/v1/email/followups/mark", json=data)

    def list_pending_followups(
        self,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List pending follow-ups.

        Args:
            limit: Maximum number of results (default: 50)
            offset: Pagination offset

        Returns:
            Dict with followups array and pagination info
        """
        return self._client.request(
            "GET",
            "/api/v1/email/followups/pending",
            params={"limit": limit, "offset": offset},
        )

    def resolve_followup(
        self,
        followup_id: str,
        resolution: str | None = None,
    ) -> dict[str, Any]:
        """
        Resolve a follow-up.

        Args:
            followup_id: ID of the follow-up to resolve
            resolution: Optional resolution notes

        Returns:
            Dict with success status
        """
        data = {}
        if resolution:
            data["resolution"] = resolution

        return self._client.request(
            "POST",
            f"/api/v1/email/followups/{followup_id}/resolve",
            json=data,
        )

    def check_replies(self) -> dict[str, Any]:
        """
        Check for replies to tracked emails.

        Returns:
            Dict with replied_count and updated followups
        """
        return self._client.request("POST", "/api/v1/email/followups/check-replies")

    # ===========================================================================
    # Snooze Management
    # ===========================================================================

    def get_snooze_suggestions(self, email_id: str) -> dict[str, Any]:
        """
        Get snooze recommendations for an email.

        Args:
            email_id: ID of the email

        Returns:
            Dict with suggested snooze times and reasons
        """
        return self._client.request("GET", f"/api/v1/email/{email_id}/snooze-suggestions")

    def snooze_email(
        self,
        email_id: str,
        snooze_until: str,
        reason: str | None = None,
    ) -> dict[str, Any]:
        """
        Apply snooze to an email.

        Args:
            email_id: ID of the email to snooze
            snooze_until: ISO datetime when snooze ends
            reason: Optional reason for snoozing

        Returns:
            Dict with snooze_id and scheduled_unsnooze
        """
        data: dict[str, Any] = {"snooze_until": snooze_until}
        if reason:
            data["reason"] = reason

        return self._client.request("POST", f"/api/v1/email/{email_id}/snooze", json=data)

    def cancel_snooze(self, email_id: str) -> dict[str, Any]:
        """
        Cancel a snooze on an email.

        Args:
            email_id: ID of the snoozed email

        Returns:
            Dict with success status
        """
        return self._client.request("DELETE", f"/api/v1/email/{email_id}/snooze")

    def list_snoozed(
        self,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List snoozed emails.

        Args:
            limit: Maximum number of results
            offset: Pagination offset

        Returns:
            Dict with snoozed emails array
        """
        return self._client.request(
            "GET",
            "/api/v1/email/snoozed",
            params={"limit": limit, "offset": offset},
        )

    # ===========================================================================
    # Email Categorization
    # ===========================================================================

    def list_categories(self) -> dict[str, Any]:
        """
        List available email categories.

        Returns:
            Dict with categories array
        """
        return self._client.request("GET", "/api/v1/email/categories")

    def submit_category_feedback(
        self,
        email_id: str,
        suggested_category: str,
        is_correct: bool,
        correct_category: str | None = None,
    ) -> dict[str, Any]:
        """
        Submit category feedback to improve categorization.

        Args:
            email_id: ID of the email
            suggested_category: The category that was suggested
            is_correct: Whether the suggestion was correct
            correct_category: The correct category if suggestion was wrong

        Returns:
            Dict with success status
        """
        data: dict[str, Any] = {
            "email_id": email_id,
            "suggested_category": suggested_category,
            "is_correct": is_correct,
        }
        if correct_category:
            data["correct_category"] = correct_category

        return self._client.request("POST", "/api/v1/email/categories/learn", json=data)


class AsyncEmailServicesAPI:
    """
    Asynchronous Email Services API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     pending = await client.email_services.list_pending_followups()
        ...     await client.email_services.mark_followup(email_id="abc123")
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # ===========================================================================
    # Follow-up Tracking
    # ===========================================================================

    async def mark_followup(
        self,
        email_id: str,
        expected_reply_by: str | None = None,
        notes: str | None = None,
    ) -> dict[str, Any]:
        """Mark an email as awaiting reply."""
        data: dict[str, Any] = {"email_id": email_id}
        if expected_reply_by:
            data["expected_reply_by"] = expected_reply_by
        if notes:
            data["notes"] = notes

        return await self._client.request("POST", "/api/v1/email/followups/mark", json=data)

    async def list_pending_followups(
        self,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List pending follow-ups."""
        return await self._client.request(
            "GET",
            "/api/v1/email/followups/pending",
            params={"limit": limit, "offset": offset},
        )

    async def resolve_followup(
        self,
        followup_id: str,
        resolution: str | None = None,
    ) -> dict[str, Any]:
        """Resolve a follow-up."""
        data = {}
        if resolution:
            data["resolution"] = resolution

        return await self._client.request(
            "POST",
            f"/api/v1/email/followups/{followup_id}/resolve",
            json=data,
        )

    async def check_replies(self) -> dict[str, Any]:
        """Check for replies to tracked emails."""
        return await self._client.request("POST", "/api/v1/email/followups/check-replies")

    # ===========================================================================
    # Snooze Management
    # ===========================================================================

    async def get_snooze_suggestions(self, email_id: str) -> dict[str, Any]:
        """Get snooze recommendations for an email."""
        return await self._client.request("GET", f"/api/v1/email/{email_id}/snooze-suggestions")

    async def snooze_email(
        self,
        email_id: str,
        snooze_until: str,
        reason: str | None = None,
    ) -> dict[str, Any]:
        """Apply snooze to an email."""
        data: dict[str, Any] = {"snooze_until": snooze_until}
        if reason:
            data["reason"] = reason

        return await self._client.request("POST", f"/api/v1/email/{email_id}/snooze", json=data)

    async def cancel_snooze(self, email_id: str) -> dict[str, Any]:
        """Cancel a snooze on an email."""
        return await self._client.request("DELETE", f"/api/v1/email/{email_id}/snooze")

    async def list_snoozed(
        self,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List snoozed emails."""
        return await self._client.request(
            "GET",
            "/api/v1/email/snoozed",
            params={"limit": limit, "offset": offset},
        )

    # ===========================================================================
    # Email Categorization
    # ===========================================================================

    async def list_categories(self) -> dict[str, Any]:
        """List available email categories."""
        return await self._client.request("GET", "/api/v1/email/categories")

    async def submit_category_feedback(
        self,
        email_id: str,
        suggested_category: str,
        is_correct: bool,
        correct_category: str | None = None,
    ) -> dict[str, Any]:
        """Submit category feedback to improve categorization."""
        data: dict[str, Any] = {
            "email_id": email_id,
            "suggested_category": suggested_category,
            "is_correct": is_correct,
        }
        if correct_category:
            data["correct_category"] = correct_category

        return await self._client.request("POST", "/api/v1/email/categories/learn", json=data)

File: sdk/typescript/src/namespaces/email-services.ts
/**
 * Email Services Namespace API
 *
 * Provides advanced email management:
 * - Follow-up tracking
 * - Snooze recommendations
 * - Category learning
 */

/**
 * Follow-up status.
 */
export type FollowUpStatus = 'awaiting' | 'overdue' | 'received' | 'resolved' | 'cancelled';

/**
 * Follow-up priority.
 */
export type FollowUpPriority = 'urgent' | 'high' | 'normal' | 'low';

/**
 * Snooze reason.
 */
export type SnoozeReason =
  | 'calendar_free'
  | 'sender_pattern'
  | 'work_hours'
  | 'priority_decay'
  | 'weekend_skip'
  | 'end_of_day'
  | 'tomorrow_morning';

/**
 * Email category.
 */
export type EmailServiceCategory =
  | 'invoices'
  | 'hr'
  | 'newsletters'
  | 'projects'
  | 'meetings'
  | 'support'
  | 'security'
  | 'receipts'
  | 'social'
  | 'personal'
  | 'uncategorized';

/**
 * Follow-up item.
 */
export interface FollowUpItem {
  followup_id: string;
  email_id: string;
  thread_id: string;
  subject: string;
  recipient: string;
  sent_at: string;
  expected_by?: string;
  status: FollowUpStatus;
  days_waiting: number;
  urgency_score?: number;
  reminder_count: number;
  resolved_at?: string;
  notes?: string;
}

/**
 * Mark follow-up request.
 */
export interface MarkFollowUpRequest {
  email_id: string;
  thread_id: string;
  subject?: string;
  recipient?: string;
  sent_at?: string;
  expected_reply_days?: number;
}

/**
 * Pending follow-ups options.
 */
export interface PendingFollowUpsOptions {
  include_resolved?: boolean;
  sort_by?: 'days_waiting' | 'expected_by' | 'urgency_score';
}

/**
 * Pending follow-ups response.
 */
export interface PendingFollowUpsResponse {
  followups: FollowUpItem[];
  total: number;
  overdue_count: number;
}

/**
 * Resolve follow-up request.
 */
export interface ResolveFollowUpRequest {
  status: 'received' | 'resolved' | 'cancelled';
  notes?: string;
}

/**
 * Resolve follow-up response.
 */
export interface ResolveFollowUpResponse {
  followup_id: string;
  status: FollowUpStatus;
  resolved_at: string;
  notes?: string;
}

/**
 * Check replies response.
 */
export interface CheckRepliesResponse {
  replied: FollowUpItem[];
  still_pending: number;
}

/**
 * Auto-detect follow-ups response.
 */
export interface AutoDetectResponse {
  detected: FollowUpItem[];
  total_detected: number;
}

/**
 * Snooze suggestion.
 */
export interface SnoozeSuggestion {
  snooze_until: string;
  label: string;
  reason: SnoozeReason;
  confidence: number;
  source?: string;
}

/**
 * Snooze suggestions options.
 */
export interface SnoozeSuggestionsOptions {
  subject: string;
  sender: string;
  priority?: string;
  max_suggestions?: number;
}

/**
 * Snooze suggestions response.
 */
export interface SnoozeSuggestionsResponse {
  email_id: string;
  suggestions: SnoozeSuggestion[];
  recommended?: SnoozeSuggestion;
}

/**
 * Snooze email response.
 */
export interface SnoozeEmailResponse {
  email_id: string;
  snooze_until: string;
  label: string;
  status: string;
}

/**
 * Snoozed email.
 */
export interface SnoozedEmail {
  email_id: string;
  snooze_until: string;
  label: string;
  snoozed_at: string;
  is_due: boolean;
}

/**
 * Snoozed emails response.
 */
export interface SnoozedEmailsResponse {
  snoozed: SnoozedEmail[];
  total: number;
  due_now: number;
}

/**
 * Process due snoozes response.
 */
export interface ProcessDueSnoozesResponse {
  processed: SnoozedEmail[];
  count: number;
}

/**
 * Category info.
 */
export interface CategoryInfo {
  id: string;
  name: string;
  description: string;
}

/**
 * Category feedback request.
 */
export interface CategoryFeedbackRequest {
  email_id: string;
  predicted_category: EmailServiceCategory;
  correct_category: EmailServiceCategory;
  email_metadata?: Record<string, unknown>;
}

/**
 * Category feedback response.
 */
export interface CategoryFeedbackResponse {
  email_id: string;
  feedback_recorded: boolean;
  predicted: EmailServiceCategory;
  correct: EmailServiceCategory;
}

/**
 * Client interface for email services operations.
 */
interface EmailServicesClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Email Services API namespace.
 *
 * Provides methods for advanced email management:
 * - Track follow-ups and awaiting replies
 * - Get smart snooze recommendations
 * - Learn from category corrections
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Mark an email for follow-up
 * const followup = await client.emailServices.markFollowUp({
 *   email_id: 'msg_123',
 *   thread_id: 'thread_456',
 *   expected_reply_days: 3,
 * });
 *
 * // Get snooze suggestions
 * const suggestions = await client.emailServices.getSnoozeSuggestions('msg_123', {
 *   subject: 'Q4 Report',
 *   sender: 'boss@company.com',
 * });
 * ```
 */
export class EmailServicesAPI {
  constructor(private client: EmailServicesClientInterface) {}

  // =========================================================================
  // Follow-up Tracking
  // =========================================================================

  /**
   * Mark an email for follow-up tracking.
   */
  async markFollowUp(request: MarkFollowUpRequest): Promise<FollowUpItem> {
    return this.client.request('POST', '/api/v1/email/followups/mark', {
      json: request as unknown as Record<string, unknown>,
    });
  }

  /**
   * Get pending follow-ups.
   */
  async getPendingFollowUps(options?: PendingFollowUpsOptions): Promise<PendingFollowUpsResponse> {
    return this.client.request('GET', '/api/v1/email/followups/pending', {
      params: options as Record<string, unknown>,
    });
  }

  /**
   * Resolve a follow-up.
   */
  async resolveFollowUp(
    followupId: string,
    status: 'received' | 'resolved' | 'cancelled',
    notes?: string
  ): Promise<ResolveFollowUpResponse> {
    return this.client.request('POST', `/api/v1/email/followups/${followupId}/resolve`, {
      json: { status, notes },
    });
  }

  /**
   * Check for replies to tracked follow-ups.
   */
  async checkReplies(): Promise<CheckRepliesResponse> {
    return this.client.request('POST', '/api/v1/email/followups/check-replies');
  }

  /**
   * Auto-detect emails needing follow-up.
   */
  async autoDetectFollowUps(daysBack?: number): Promise<AutoDetectResponse> {
    return this.client.request('POST', '/api/v1/email/followups/auto-detect', {
      json: daysBack ? { days_back: daysBack } : undefined,
    });
  }

  // =========================================================================
  // Snooze Management
  // =========================================================================

  /**
   * Get snooze suggestions for an email.
   */
  async getSnoozeSuggestions(
    emailId: string,
    options: SnoozeSuggestionsOptions
  ): Promise<SnoozeSuggestionsResponse> {
    return this.client.request('GET', `/api/v1/email/${emailId}/snooze-suggestions`, {
      params: options as unknown as Record<string, unknown>,
    });
  }

  /**
   * Snooze an email.
   */
  async snoozeEmail(
    emailId: string,
    snoozeUntil: string,
    label?: string
  ): Promise<SnoozeEmailResponse> {
    return this.client.request('POST', `/api/v1/email/${emailId}/snooze`, {
      json: { snooze_until: snoozeUntil, label },
    });
  }

  /**
   * Unsnooze an email.
   */
  async unsnoozeEmail(emailId: string): Promise<{ email_id: string; status: string }> {
    return this.client.request('DELETE', `/api/v1/email/${emailId}/snooze`);
  }

  /**
   * Get all snoozed emails.
   */
  async getSnoozedEmails(): Promise<SnoozedEmailsResponse> {
    return this.client.request('GET', '/api/v1/email/snoozed');
  }

  /**
   * Process due snoozes (un-snooze emails that are due).
   */
  async processDueSnoozes(): Promise<ProcessDueSnoozesResponse> {
    return this.client.request('POST', '/api/v1/email/snooze/process-due');
  }

  // =========================================================================
  // Category Management
  // =========================================================================

  /**
   * Get available email categories.
   */
  async getCategories(): Promise<{ categories: CategoryInfo[] }> {
    return this.client.request('GET', '/api/v1/email/categories');
  }

  /**
   * Submit category feedback for learning.
   */
  async learnCategory(feedback: CategoryFeedbackRequest): Promise<CategoryFeedbackResponse> {
    return this.client.request('POST', '/api/v1/email/categories/learn', {
      json: feedback as unknown as Record<string, unknown>,
    });
  }
}
